src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.MethodParametersInfo.cs

```diff
@@ -0,0 +1,305 @@
+
+namespace Microsoft.CodeAnalysis.InlineMethod
+{
+    internal abstract partial class AbstractInlineMethodRefactoringProvider<TInvocationSyntax, TExpressionSyntax, TMethodDeclarationSyntax, TStatementSyntax>
+    {
+        /// <summary>
+        /// Information about the callee method parameters to compute <see cref="InlineMethodContext"/>.
+        /// </summary>
+        private class MethodParametersInfo
+        {
+            /// <summary>
+            /// Parameters map to identifier argument's name. The identifier from Caller will be used to replace
+            /// the mapping callee's parameter.
+            /// Example:
+            /// Before:
+            /// void Caller(int i, int j, bool[] k)
+            /// {
+            ///     Callee(i, j, k);
+            /// }
+            /// void Callee(int a, int b, params bool[] c)
+            /// {
+            ///     DoSomething(a, b, c);
+            /// }
+            /// After:
+            /// void Caller(int i, int j, bool[] k)
+            /// {
+            ///     DoSomething(i, j, k);
+            /// }
+            /// void Callee(int a, int b, params bool[] c)
+            /// {
+            ///     DoSomething(a, b, c);
+            /// }
+            /// </summary>
+            public ImmutableDictionary<IParameterSymbol, string> ParametersWithIdentifierArgument { get; }
+
+            /// <summary>
+            /// Parameters map to variable declaration argument's name.
+            /// This is only used for C# to support the 'out' variable declaration. For VB it should always be empty.
+            /// Before:
+            /// void Caller()
+            /// {
+            ///     Callee(out var x);
+            /// }
+            /// void Callee(out int i) => i = 100;
+            ///
+            /// After:
+            /// void Caller()
+            /// {
+            ///     int x = 100;
+            /// }
+            /// void Callee(out int i) => i = 100;
+            /// </summary>
+            public ImmutableArray<(IParameterSymbol parameterSymbol, string)> ParametersWithVariableDeclarationArgument { get; }
+
+            /// <summary>
+            /// Operations that represent Parameter has argument but the argument is not identifier or literal.
+            /// For these parameters they are considered to be put into a declaration statement after inlining.
+            /// Note: params array could maps to multiple/zero arguments.
+            /// Example:
+            /// Before:
+            /// void Caller(bool x)
+            /// {
+            ///     Callee(Foo(), x ? Foo() : Bar())
+            /// }
+            /// void Callee(int a, int b)
+            /// {
+            ///     DoSomething(a, b);
+            /// }
+            /// After:
+            /// void Caller(bool x)
+            /// {
+            ///     int a = Foo();
+            ///     int b = x ? Foo() : Bar();
+            ///     DoSomething(a, b);
+            /// }
```

Comment:
so this def works... but it seems unfortunately onerous in the case where the parameter is only used once.  ie.. my expectation of the above would be that we would convert it to:

```
DoSomething(Foo(), x ? Foo() : Bar());
```

However, if the parameter were used multiple times, or used in a different order, then i might expect the above.  

Effectively if the args are passed in some order and used in that same order, it seems like we can just preserve that.

thoughts?

--

note: i do see taht this strictly preserves semantics (which is good).  i'm just wondering if we should be a little more lenient for cases that appear safe, even if they technically *might* change semantics...

I'm not sure.  What do you think?

Reply:
I used to think this when I implemented this part of code and at that time I just choose the safest way. :)
I can think that 95 percent of the users might want that, and the remain 5% users doesn't want because they are expecting the original semantics. (And they might need to debug the code to figure out why because I feel the semantics change is not very obvious)
Generally I think this would be a good if our discpline is that a little semantics changes (errors) are allowed. :)

Reply:
OK as we discussed offline, I have added implementation for this part.
Enable this scenario if the parameter is only 'read' once.